File: packages/node/src/diagnose/store.ts
/**
 * Run persistence (SPEC.md §4): .tryaii/diagnose/<run-id>/ + latest pointer.
 *
 * The `latest` pointer is a plain text file (never a symlink — Windows).
 * Run ids sort lexicographically, so "previous run" is a plain sorted
 * lookup. All files are written with \n newlines for cross-platform byte
 * parity. Mirrors diagnose/store.py.
 */

import { existsSync, mkdirSync, readdirSync, readFileSync, statSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

export const LATEST_FILE = 'latest';

function dump(obj: unknown): string {
  return JSON.stringify(obj, null, 2) + '\n';
}

/** Persist one run; returns {name: path} of everything written. */
export function writeRun(
  outDir: string,
  inventoryData: unknown,
  findings: Record<string, unknown>,
): Record<string, string> {
  const runId = findings.run_id as string;
  const runDir = join(outDir, runId);
  mkdirSync(runDir, { recursive: true });

  const meta = {
    run_id: runId,
    generated_at: findings.generated_at,
    tool: findings.tool,
  };
  const paths = {
    inventory: join(runDir, 'inventory.json'),
    findings: join(runDir, 'findings.json'),
    meta: join(runDir, 'meta.json'),
    latest: join(outDir, LATEST_FILE),
  };
  writeFileSync(paths.inventory, dump(inventoryData), 'utf-8');
  writeFileSync(paths.findings, dump(findings), 'utf-8');
  writeFileSync(paths.meta, dump(meta), 'utf-8');
  writeFileSync(paths.latest, runId + '\n', 'utf-8');
  return paths;
}

/** Run dirs (containing a findings.json), sorted ascending. */
export function listRunIds(outDir: string): string[] {
  if (!existsSync(outDir) || !statSync(outDir).isDirectory()) return [];
  return readdirSync(outDir)
    .filter((name) => {
      const p = join(outDir, name);
      return statSync(p).isDirectory() && existsSync(join(p, 'findings.json'));
    })
    .sort();
}

/** The pointer file's run id, falling back to the newest run dir. */
export function latestRunId(outDir: string): string | null {
  const pointer = join(outDir, LATEST_FILE);
  if (existsSync(pointer) && statSync(pointer).isFile()) {
    const runId = readFileSync(pointer, 'utf-8').trim();
    if (runId && existsSync(join(outDir, runId, 'findings.json'))) return runId;
  }
  const runs = listRunIds(outDir);
  return runs.length ? runs[runs.length - 1] : null;
}

/** The run immediately before `runId` in lexicographic order. */
export function previousRunId(outDir: string, runId: string): string | null {
  const earlier = listRunIds(outDir).filter((r) => r < runId);
  return earlier.length ? earlier[earlier.length - 1] : null;
}

export function loadRunFindings(outDir: string, runId: string): Record<string, unknown> {
  return JSON.parse(readFileSync(join(outDir, runId, 'findings.json'), 'utf-8')) as Record<
    string,
    unknown
  >;
}

File: packages/python/tryaii/diagnose/store.py
"""Run persistence (SPEC.md §4): .tryaii/diagnose/<run-id>/ + latest pointer.

The `latest` pointer is a plain text file (never a symlink — Windows).
Run ids sort lexicographically, so "previous run" is a plain sorted lookup.
All files are written with \\n newlines for cross-platform byte parity.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Optional

LATEST_FILE = "latest"


def _dump(obj) -> str:
    return json.dumps(obj, indent=2, ensure_ascii=False) + "\n"


def _write(path: Path, text: str) -> None:
    with open(path, "w", encoding="utf-8", newline="\n") as fh:
        fh.write(text)


def write_run(out_dir: Path, inventory_data, findings: dict) -> dict:
    """Persist one run; returns {name: Path} of everything written."""
    out_dir = Path(out_dir)
    run_id = findings["run_id"]
    run_dir = out_dir / run_id
    run_dir.mkdir(parents=True, exist_ok=True)

    meta = {"run_id": run_id,
            "generated_at": findings["generated_at"],
            "tool": findings["tool"]}
    paths = {
        "inventory": run_dir / "inventory.json",
        "findings": run_dir / "findings.json",
        "meta": run_dir / "meta.json",
        "latest": out_dir / LATEST_FILE,
    }
    _write(paths["inventory"], _dump(inventory_data))
    _write(paths["findings"], _dump(findings))
    _write(paths["meta"], _dump(meta))
    _write(paths["latest"], run_id + "\n")
    return paths


def list_run_ids(out_dir: Path) -> list[str]:
    """Run dirs (containing a findings.json), sorted ascending."""
    out_dir = Path(out_dir)
    if not out_dir.is_dir():
        return []
    return sorted(
        p.name for p in out_dir.iterdir()
        if p.is_dir() and (p / "findings.json").is_file())


def latest_run_id(out_dir: Path) -> Optional[str]:
    """The pointer file's run id, falling back to the newest run dir."""
    pointer = Path(out_dir) / LATEST_FILE
    if pointer.is_file():
        run_id = pointer.read_text(encoding="utf-8").strip()
        if run_id and (Path(out_dir) / run_id / "findings.json").is_file():
            return run_id
    runs = list_run_ids(out_dir)
    return runs[-1] if runs else None


def previous_run_id(out_dir: Path, run_id: str) -> Optional[str]:
    """The run immediately before `run_id` in lexicographic order."""
    earlier = [r for r in list_run_ids(out_dir) if r < run_id]
    return earlier[-1] if earlier else None


def load_run_findings(out_dir: Path, run_id: str) -> dict:
    path = Path(out_dir) / run_id / "findings.json"
    return json.loads(path.read_text(encoding="utf-8"))
